### PR TITLE
Allow Mesh-related queue phase systems to parallelize

### PR DESCRIPTION
**Pull request by james7132**
# Objective
Partially addresses #3548. `queue_shadows` and `queue_material_meshes` cannot parallelize because of the `ResMut<RenderMeshInstances>` parameter for `queue_material_meshes`.

## Solution
Change the `material_bind_group` field to use atomics instead of needing full mutable access. Change the `ResMut` to a `Res`, which should allow both sets of systems to parallelize without issue.

## Performance
Tested against `many_foxes`, this has a significant improvement over the entire render schedule. (Yellow is this PR, red is main)
image

The use of atomics does seem to have a negative effect on `queue_material_meshes` (roughly a 8.29% increase in time spent in the system).
image

`queue_shadows` seems to be ever so slightly slower (1.6% more time spent) in the system.
image

`batch_and_prepare_render_phase` seems to be a mix, but overall seems to be slightly *faster* by about 5%.
image

### CODE REVIEW

File: crates/bevy_pbr/src/lightmap/mod.rs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-//! Lightmaps, baked lighting textures that can be applied at runtime to provide
+//! Lightmaps, bailed lighting textures that can be applied at runtime to provide
```
Comment (pcwalton) [inline crates/bevy_pbr/src/lightmap/mod.rs:1]: Typo

File: crates/bevy_pbr/src/material.rs [hunk 2]
```diff
@@ -795,6 +796,24 @@ pub struct PreparedMaterial<T: Material> {
 #[derive(Component, Clone, Copy, Default, PartialEq, Eq, Deref, DerefMut)]
 pub struct MaterialBindGroupId(Option<BindGroupId>);
 
+#[derive(Default)]
+pub struct AtomicMaterialBindGroupId(AtomicU32);
+
+impl AtomicMaterialBindGroupId {
+    pub fn set(&self, id: MaterialBindGroupId) {
+        let id = if let Some(id) = id.0 {
+            core::num::NonZeroU32::from(id).get()
```
Comment (pcwalton) [inline crates/bevy_pbr/src/material.rs:805]: Can just be `NonZeroU32` given that you imported it already

**Review by JMS55 — APPROVED** [review]

File: crates/bevy_pbr/src/lightmap/mod.rs (hunk 1 above)
Comment (james7132) [reply crates/bevy_pbr/src/lightmap/mod.rs:1]: Not sure why that happened. Nice catch.

File: crates/bevy_pbr/src/material.rs (hunk 2 above)
Comment (james7132) [reply crates/bevy_pbr/src/material.rs]: Done.

**Comment by hymm** [comment]
Does this have a significant effect on single threaded perf?

**Review by superdump — APPROVED** [review]

**Comment by superdump** [comment]
Merge conflict. Code looks good. Could you test with a heavier load like `many_cubes`?

**Review by IceSentry — APPROVED** [review]
LGTM assuming conflicts are fixed

**Comment by james7132** [comment]
> Merge conflict. Code looks good. Could you test with a heavier single-material-type no-contention load like `many_cubes`?

Checked against main and this PR. There seems to be next to zero impact on the average time spent, which makes sense uncontended atomics on x86 machines is essentially free.

**Comment by james7132** [comment]
> Does this have a significant effect on single threaded perf?

Most modern platforms (with maybe the exception of wasm) treat all load/store operations as atomic, and so long as there is no contention on the same cache line, it's identical to single threaded performance. The only thing you might be missing out on is the compiler reordering your operations to be more optimal, which isn't a particular concern for these systems.

**Pull request merged**